### PR TITLE
custom distillation units and qubit error definitions

### PR DESCRIPTION
**Pull request by ivanbasov**

### CODE REVIEW

File: azure-quantum/azure/quantum/target/microsoft/target.py [hunk 1]
```diff
@@ -32,6 +33,32 @@ def _check_error_rate(name, value):
...
+
+    if isinstance(value, float):
+        _check_error_rate(name, value)
+        return
+
+    if not isinstance(value, MeasurementErrorRate):
+        raise ValueError(f"{name} must be either a float or MeasurementErrorRate with two fields: 'process' and 'readout'")
+
+    if value.process is None:
+            raise LookupError("process must be set")
```
Comment (shenyingjun) [inline azure-quantum/azure/quantum/target/microsoft/target.py:48]: nit: indent seems to be wrong. same as below

File: azure-quantum/azure/quantum/target/microsoft/target.py [hunk 2]
```diff
@@ -156,6 +255,31 @@ def as_dict(self, validate=True) -> Dict[str, Any]:
...
+                if specification.physical_qubit_specification is not None:
+                    physical_qubit_specification_dict = specification.physical_qubit_specification.as_dict(validate)
+                    if len(physical_qubit_specification_dict) != 0:
+                        specification_dict["physicalQubitSpecification"] = physical_qubit_specification_dict
+
+                if specification.logical_qubit_specification is not None:
+                    logical_qubit_specification_dict = specification.logical_qubit_specification.as_dict(validate)
+                    if len(logical_qubit_specification_dict) != 0:
+                        specification_dict["logicalQubitSpecification"] = logical_qubit_specification_dict
+
```
Comment (shenyingjun) [inline azure-quantum/azure/quantum/target/microsoft/target.py:273]: nit: extra line

File: azure-quantum/azure/quantum/target/microsoft/target.py [hunk 3]
```diff
@@ -156,6 +255,31 @@ def as_dict(self, validate=True) -> Dict[str, Any]:
...
+
+                if specification.logical_qubit_specification_first_round_override is not None:
+                    logical_qubit_specification_first_round_override_dict = specification.logical_qubit_specification_first_round_override.as_dict(validate)
+                    if len(logical_qubit_specification_first_round_override_dict) != 0:
+                        specification_dict["logicalQubitSpecificationFirstRoundOverride"] = logical_qubit_specification_first_round_override_dict
+
+                result["distillationUnitSpecifications"].append(
```
Comment (shenyingjun) [inline azure-quantum/azure/quantum/target/microsoft/target.py:280]: nit: result["distillationUnitSpecifications"].append(specification_dict)

**Review by shenyingjun — APPROVED** [review]

File: azure-quantum/azure/quantum/target/microsoft/target.py [hunk 4]
```diff
@@ -156,6 +255,29 @@ def as_dict(self, validate=True) -> Dict[str, Any]:
         if len(qec_scheme) != 0:
             result["qecScheme"] = qec_scheme
 
+        for specification in self.distillation_unit_specifications:
```
Comment (msoeken) [inline azure-quantum/azure/quantum/target/microsoft/target.py:258]: This logic fits better into the `DistillationUnitSpecification` class.

**Review by msoeken — CHANGES_REQUESTED** [review]
Thanks, @ivanbasov , looks good to me, but I have some comments.

File: azure-quantum/tests/unit/test_microsoft_qc.py
[the remark this answers is not in the document]
Comment (ivanbasov) [reply azure-quantum/tests/unit/test_microsoft_qc.py:1]: Thank you, @msoeken!
I believe `test_estimator_batching_job` I modified is a live test. It looks like the only success path live test for the RE. Do you think we should split it into several tests? Or should we have all success scenarios combined at a single one?

File: azure-quantum/azure/quantum/target/microsoft/target.py [hunk 5]
```diff
@@ -207,6 +203,33 @@ def post_validation(self, result):
         if self.logical_qubit_specification_first_round_override is not None:
             self.logical_qubit_specification_first_round_override.post_validation(result)
 
+    def as_dict_custom(self, validate):
```
Comment (msoeken) [inline azure-quantum/azure/quantum/target/microsoft/target.py:206]: Could we override `as_dict` and then call `super.as_dict` within?

**Review by msoeken — APPROVED** [review]

File: azure-quantum/tests/unit/test_microsoft_qc.py [hunk 6]
```diff
@@ -231,6 +255,22 @@ def test_estimator_params_validation_missing_fields(self):
...
+        params.qubit_params.name = QubitParams.GATE_NS_E3
+        params.qubit_params.instruction_set = "gate_based"
+        params.qubit_params.t_gate_error_rate = 0.03
+        params.qubit_params.t_gate_time = "10 ns"
+        params.qubit_params.idle_error_rate = 0.02
+        params.one_qubit_measurement_error_rate = 0.01
+        params.qubit_params.two_qubit_joint_measurement_error_rate = \
+        MeasurementErrorRate(process = 0.02, readout = 0.03)
+
+        params.as_dict()
```
Comment (msoeken) [inline azure-quantum/tests/unit/test_microsoft_qc.py:271]: You should compare this to an expected value. You will notice that the `MeasurementErrorRate` is not correctly serialized.

**Review by msoeken — CHANGES_REQUESTED** [review]
There is still a problem with serializing `MeasurementErrorRate` into a dictionary.

**Review by msoeken — APPROVED** [review]

**Pull request merged**